modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1976,27 +2084,71 @@ def _execute_arrow(self):
         pyarrow.Table
             The resulting table.
         """
-        if isinstance(self._op, FrameNode):
-            if self._partitions.size == 0:
-                return pyarrow.Table.from_pandas(pd.DataFrame({}))
+        result = None
+        stack = [self]
+
+        while stack:
+            frame = stack.pop()
+
+            if callable(frame):
+                frame()
+            elif isinstance(frame._op, FrameNode):
+                if frame._partitions.size == 0:
+                    result = pyarrow.Table.from_pandas(pd.DataFrame({}))
+                else:
+                    assert frame._partitions.size == 1
+                    result = frame._partitions[0][0].get()
+            elif isinstance(frame._op, MaskNode):
+
+                def slice(positions=frame._op.row_positions):
+                    nonlocal result
+                    result = self._arrow_row_slice(result, positions)
+
+                stack.append(slice)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, TransformNode):
+
+                def select(exprs=frame._op.exprs):
+                    nonlocal result
+                    result = self._arrow_select(result, exprs)
+
+                stack.append(select)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, UnionNode):
+
+                def union(op=frame._op, tables={}, input=iter(frame._op.input)):
+                    nonlocal result
+
+                    if (i := next(input, None)) is None:
+                        result = self._union_all_arrow(
+                            op.input, op.join, op.sort, op.ignore_index, tables
+                        )
+                    else:
+
+                        def add_result(f=i):
+                            tables[f] = result
+
+                        stack.append(frame if callable(frame) else union)
```

Comment:
how can the frame be callable here? we just accessed its `._op` field few lines above

Reply:
We accessed `._op` in the function arguments declaration. When the function is created, `frame` is of type `UnionNode` and we have the instance check here - https://github.com/AndreyPavlenko/modin/blob/issue-5889/modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py#L2117. The function is called here https://github.com/AndreyPavlenko/modin/blob/issue-5889/modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py#L2094 and, when the function is called, `frame` is a reference to this function and we add this function to the stack again.